core/src/main/java/jenkins/fingerprints/FileFingerprintStorage.java

```diff
@@ -0,0 +1,188 @@
+            if (!(loaded instanceof Fingerprint)) {
+                throw new IOException("Unexpected Fingerprint type. Expected " + Fingerprint.class + " or subclass but got "
+                        + (loaded != null ? loaded.getClass() : "null"));
+            }
+            Fingerprint f = (Fingerprint) loaded;
+            return f;
+        } catch (IOException e) {
+            if(file.exists() && file.length()==0) {
+                // Despite the use of AtomicFile, there are reports indicating that people often see
+                // empty XML file, presumably either due to file system corruption (perhaps by sudden
+                // power loss, etc.) or abnormal program termination.
+                // generally we don't want to wipe out user data just because we can't load it,
+                // but if the file size is 0, which is what's reported in HUDSON-2012, then it seems
+                // like recovering it silently by deleting the file is not a bad idea.
+                logger.log(Level.WARNING, "Size zero fingerprint. Disk corruption? {0}", configFile);
+                file.delete();
+                return null;
+            }
+            String parseError = messageOfParseException(e);
+            if (parseError != null) {
+                logger.log(Level.WARNING, "Malformed XML in {0}: {1}", new Object[] {configFile, parseError});
+                file.delete();
+                return null;
+            }
+            logger.log(Level.WARNING, "Failed to load "+configFile,e);
+            throw e;
+        }
+    }
+
+    public synchronized void save(Fingerprint fp) throws IOException {
+        File file = getFingerprintFile(fp.getHashString());
+        save(fp, file);
+        SaveableListener.fireOnChange(fp, getConfigFile(file));
+    }
+
+    public static void save(Fingerprint fp, File file) throws IOException {
+        if (fp.getPersistedFacets().isEmpty()) {
+            file.getParentFile().mkdirs();
+            // JENKINS-16301: fast path for the common case.
+            AtomicFileWriter afw = new AtomicFileWriter(file);
+            try {
+                PrintWriter w = new PrintWriter(afw);
+                w.println("<?xml version='1.1' encoding='UTF-8'?>");
+                w.println("<fingerprint>");
+                w.print("  <timestamp>");
+                w.print(DATE_CONVERTER.toString(fp.getTimestamp()));
+                w.println("</timestamp>");
+                if (fp.getOriginal() != null) {
+                    w.println("  <original>");
+                    w.print("    <name>");
+                    w.print(Util.xmlEscape(fp.getOriginal().getName()));
+                    w.println("</name>");
+                    w.print("    <number>");
+                    w.print(fp.getOriginal().getNumber());
+                    w.println("</number>");
+                    w.println("  </original>");
+                }
+                w.print("  <md5sum>");
```

Comment:
This tag might become a source of confusion later if the fingerprint size is expanded. Anyway, we can worry about it when we get there.

```suggestion
                // TODO(oleg_nenashev): Consider renaming the field: https://issues.jenkins-ci.org/browse/JENKINS-25808
                w.print("  <md5sum>");
```

Reply:
I absolutely agree :) TODO added